### PR TITLE
Add checkpoint support to stock implementations

### PR DESCRIPTION
**Pull request by HerbertJordan**
This PR adds checkpointing support to all stock implementations.

While implementing the tests it was realized that the `Restore` operation of a checkpoint can not be a member function of the targeted data structure itself. This would require the ability to open a corrupted data structure to be able to restore it. Supporting this would be very complicated, and it is not needed.

Instead, the `Restore` functionality was moved into its own interface, independent of the data structure it is targeting. This way, a corrupted data structure can be restored before attempting to reopen the data structure.

This is the second step toward supporting checkpointing of archives (see #946)

### CODE REVIEW

**Review by petr-hanzl — COMMENTED** [review]
I think the filenames should be `constants` as I can see a lot of usage solely in this PR.

File: go/backend/stock/file/file.go [hunk 1]
```diff
@@ -214,17 +228,27 @@ func verifyStackInternal[I stock.Index](meta metadata, freelistfile utils.OsFile
...
 	return
 }
 
+func getCommittedCheckpointFile(directory string) string {
+	return filepath.Join(directory, "committed.json")
+}
+
+func getPendingCheckpointFile(directory string) string {
+	return filepath.Join(directory, "pending.json")
+}
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/stock/file/file.go:243]: Done

File: go/backend/stock/memory/memory.go [hunk 2]
```diff
@@ -251,13 +281,103 @@ func (s *inMemoryStock[I, V]) Close() error {
...
+	}
+	return meta, nil
+}
+
+func writeCheckpointMetaData(path string, data checkpointData) error {
+	meta, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(path, meta, 0600)
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/stock/memory/memory.go]: I factored out these type of functions into a common utility library #978 to get rid of this kind of redundancy.

[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/stock/memory/memory.go]: I replaced those functions with generic utility functions.

File: go/backend/stock/memory/memory.go [hunk 3]
```diff
@@ -37,13 +45,13 @@ func OpenStock[I stock.Index, V any](encoder stock.ValueEncoder[V], directory st
...
 
-	// Create the direcory if needed.
+	// Create the directory if needed.
 	if err := os.MkdirAll(directory, 0700); err != nil {
 		return nil, err
 	}
 
 	// Test whether a meta file exists in this directory.
-	metafile := directory + "/meta.json"
+	metafile := filepath.Join(directory, "meta.json")
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:54]:
```suggestion
	metafile := filepath.Join(directory, fileNameMetadata)
```

File: go/backend/stock/memory/memory.go [hunk 4]
```diff
@@ -74,7 +82,7 @@ func OpenStock[I stock.Index, V any](encoder stock.ValueEncoder[V], directory st
 
 	// Load list of values.
 	{
-		valuefile := directory + "/values.dat"
+		valuefile := filepath.Join(directory, "values.dat")
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:85]:
```suggestion
		valuefile := filepath.Join(directory, fileNameValues)
```

File: go/backend/stock/memory/memory.go [hunk 5]
```diff
@@ -102,7 +110,7 @@ func OpenStock[I stock.Index, V any](encoder stock.ValueEncoder[V], directory st
 
 	// Load freelist.
 	{
-		freelistfile := directory + "/freelist.dat"
+		freelistfile := filepath.Join(directory, "freelist.dat")
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:113]:
```suggestion
		freelistfile := filepath.Join(directory, fileNameFreeList)
```

File: go/backend/stock/memory/memory.go [hunk 6]
```diff
@@ -188,25 +207,37 @@ func (s *inMemoryStock[I, V]) GetMemoryFootprint() *common.MemoryFootprint {
...
+		FreeListLength:                len(s.freeList),
+		LastCheckpoint:                s.checkpoint,
+		LastCheckpointValueListLength: int(s.checkpointValueListLength),
+		LastCheckpointFreeListLength:  int(s.checkpointFreeListLength),
 	})
 	if err != nil {
 		return err
 	}
-	if err := os.WriteFile(s.directory+"/meta.json", metadata, 0600); err != nil {
+	if err := os.WriteFile(filepath.Join(dir, "meta.json"), metadata, 0600); err != nil {
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:235]:
```suggestion
	if err := os.WriteFile(filepath.Join(dir, fileNameMetadata), metadata, 0600); err != nil {
```

File: go/backend/stock/memory/memory.go [hunk 7]
```diff
@@ -188,25 +207,37 @@ func (s *inMemoryStock[I, V]) GetMemoryFootprint() *common.MemoryFootprint {
...
 		return err
 	}
 
 	// Write list of values.
-	if f, err := os.Create(s.directory + "/values.dat"); err != nil {
+	if f, err := os.Create(filepath.Join(dir, "values.dat")); err != nil {
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:240]:
```suggestion
	if f, err := os.Create(filepath.Join(dir, fileNameValues)); err != nil {
```

File: go/backend/stock/memory/memory.go [hunk 8]
```diff
@@ -226,7 +257,7 @@ func (s *inMemoryStock[I, V]) Flush() error {
 	}
 
 	// Write free list.
-	if f, err := os.Create(s.directory + "/freelist.dat"); err != nil {
+	if f, err := os.Create(filepath.Join(dir, "freelist.dat")); err != nil {
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:260]:
```suggestion
	if f, err := os.Create(filepath.Join(dir, fileNameFreeList)); err != nil {
```

File: go/backend/stock/memory/memory.go [hunk 9]
```diff
@@ -251,13 +282,83 @@ func (s *inMemoryStock[I, V]) Close() error {
...
+
+func (s *inMemoryStock[I, V]) Abort(commit checkpoint.Checkpoint) error {
+	if s.checkpoint+1 != commit {
+		return fmt.Errorf("unable to abort checkpoint %d, expected %d", commit, s.checkpoint+1)
+	}
+	s.checkpointValueListLength = s.backupCheckpointValueListLength
+	s.checkpointFreeListLength = s.backupCheckpointFreeListLength
+	prepareFile := filepath.Join(s.directory, "prepare.json")
+	return os.Remove(prepareFile)
+}
```
Comment (petr-hanzl) [inline go/backend/stock/memory/memory.go:338]: Lets keep the code consistent and also create constants for `prepare.json` and `commit.json`

**Review by petr-hanzl — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**